library/bignum.c

```diff
@@ -1689,7 +1689,9 @@ int mbedtls_mpi_exp_mod(mbedtls_mpi *X, const mbedtls_mpi *A,
      * Correct for negative A.
      */
     if (A->s == -1 && (E->p[0] & 1) != 0) {
```

Comment:
I guess something like this?
```
/* If A < 0 then the result may be negative (if E is odd and the result is nonzero). We don't mind leaking the sign of A but we don't want to leak the low bit of E so we make sure to access E only through constant-time functions, to avoid the compiler potentially reordering checks and generating a conditional jump on the low bit of E. */
if (A->s == -1) {
    mbedtls_ct_condition_t is_x_non_zero = mbedtls_mpi_core_check_zero_ct(X->p, X->n);
    mbedtls_ct_condition_t is_e_odd = mbedtls_ct_uint_ne(E->p[0] & 1, 0);
    mbedtls_ct_condition_t is_negative_result = mbedtls_ct_bool_and(is_x_non_zero, is_e_odd);
    X->s = mbedtls_ct_uint_if(is_negative_result, -1, 1);
}
```

Reply:
We are leaking the high order bits of E at the sanity check phase in the beginning of the function. I am not sure it is worth trying to protect the last bit of E at this point.

Reply:
> We are leaking the high order bits of E at the sanity check phase in the beginning of the function.

I'm uneasy about that as well, but we can improve in a follow-up.